File: leetCode-swift/leetCode-swift/leetCode-0001.swift
//
//  leetCode-0001.swift
//  leetCode-swift
//
//  Created by liyoucheng on 2018/11/21.
//  Copyright © 2018年 Giga. All rights reserved.
//

import Foundation

func twoSum(_ nums: [Int], _ target: Int) -> [Int] {
    var dict = [Int: Int]()
    
    for (i, num) in nums.enumerated() {
        if let lastIndex = dict[target - num] {
            return [lastIndex, i]
        }
        else {
            dict[num] = i;
        }
    }
    
    return []
}

func test_0001() {
    let nums = [ 2, 7, 11, 15 ]
    let target = 9
    let ret = twoSum(nums, target)
    print(ret)
}

File: leetCode-c/leetCode-c/LeetCode/leetCode-0001/leetCode-0001.cpp
//
//  leetCode-0001.cpp
//  leetCode-c
//
//  Created by liyoucheng on 2019/1/20.
//  Copyright © 2019年 Giga. All rights reserved.
//

#include "leetCode-0001.hpp"

#include <iostream>
#include <queue>
#include <unordered_map>

using namespace std;

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        
        unordered_map<int, int> m;
        vector<int> res;
        
        // 使用一个HashMap，来建立数字和其坐标位置之间的映射
        for (int i = 0; i < nums.size(); ++i) {
            
            m[nums[i]] = i;
        }
        
        for (int i = 0; i < nums.size(); ++i) {
            
            // 用target减去遍历到的数字，就是另一个需要的数字了
            int t = target - nums[i];
            // 在HashMap中查找其是否存在
            if (m.count(t) && m[t] != i) {
                
                res.push_back(i);
                res.push_back(m[t]);
                break;
            }
        }
        
        return res;
    }
};
